core/src/consensus.rs

```diff
@@ -569,13 +569,36 @@ impl Tower {
                     return SwitchForkDecision::SameFork;
                 }
 
-                assert!(
-                    !last_vote_ancestors.contains(&switch_slot),
-                    "Should never consider switching to slot ({}), which is ancestors({:?}) of last vote: {}",
-                    switch_slot,
-                    last_vote_ancestors,
-                    last_voted_slot
-                );
+                if last_vote_ancestors.contains(&switch_slot) {
+                    if !self.is_stray_last_vote() {
+                        panic!(
+                            "Should never consider switching to slot ({}), which is ancestors({:?}) of last vote: {}",
+                            switch_slot,
+                            last_vote_ancestors,
+                            last_voted_slot
+                        );
+                    } else {
+                        // This peculiar corner handling is needed mainly for a tower which is newer than
+                        // blockstore. (Yeah, we tolerate it for ease of maintaining validator by operators)
+                        // This condition could be introduced by manual ledger mishandling,
+                        // validator SEGV, OS/HW crash, or plain No Free Space FS error.
+                        // When we're in this clause, it basically means validator is badly running
+                        // with a future tower and RE-CREATING a block for one of past slots as the leader
```

Comment:
For my clarification, does the case described here look something like:

Validator `A` is generated blocks for slots `1` through `5`, but only voted on `1` through `4` before it crashed.

`Tower: 1 -> 2 -> 3 -> 4`

Blockstore is corrupted deleted, validator starts up again from snapshot of slot `1`.

After receiving the blocks from the network, the validator freezes bank `3`, it cannot vote on it due to lockouts.
When the validator creates bank `4`, now bank forks has the ancestors of bank 4, so `ancestors(4).contains(3)` will return true.

Now before bank `4` can be frozen, heaviest for choice rule will try to pick `3` again, this time `ancestors(4).contains(3)` returns true so the panic occurs.

If this is the case, luckily the current `is_locked_out` check: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1578,  specifically this check:
https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L474-L476, should at least make sure the vote doesn't happen.

Unfortunately, I don't think this prevents creating a duplicate block. For instance in the example above, the validator previously created a block `5`, but has no idea that it's done so because the ledger is gone. Then even without voting for block `4`, PoH will reset to `4`:  https://github.com/ryoqun/solana/blob/master/core/src/replay_stage.rs#L529 because it's the heaviest bank, and the leader may generate another version of block `5`.

Reply:
@carllin yeah. all are true! So glad you understand correctly. And thanks for writing explation back to me for me, taking my work. xD